model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaEntityVersionListener.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2022. Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.storage.jpa.hibernate.listeners;
+
+import org.hibernate.HibernateException;
+import org.hibernate.event.spi.PreDeleteEvent;
+import org.hibernate.event.spi.PreDeleteEventListener;
+import org.hibernate.event.spi.PreInsertEvent;
+import org.hibernate.event.spi.PreInsertEventListener;
+import org.hibernate.event.spi.PreUpdateEvent;
+import org.hibernate.event.spi.PreUpdateEventListener;
+import org.keycloak.models.map.storage.jpa.JpaChildEntity;
+import org.keycloak.models.map.storage.jpa.JpaRootEntity;
+
+/**
+ * Listen on changes on child- and root entities and updates the current entity version of the root.
+ *
+ * This support a multiple level parent-child relationship, where only the upmost parent is locked.
+ */
+public class JpaEntityVersionListener implements PreInsertEventListener, PreDeleteEventListener, PreUpdateEventListener {
+
+    public static final JpaEntityVersionListener INSTANCE = new JpaEntityVersionListener();
+
+    /**
+     * Check if the entity is a child with a parent and force optimistic locking increment on the upmost parent.
```

Comment:
nitpick: this javadoc was prob copied over from the other listener and doesn't correspond to what the method does.

Reply:
Thanks, I updated that. The danger of writing comments: they might get out of date. I try to write comment why I do something or the intent, not necessarily how I do things. Think I got a bit off-track here. JavaDoc updated, please re-review.